### PR TITLE
Add address lookup table program

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Users and protocols need a way to store addresses on-chain for transaction v2 address lookups.

#### Summary of Changes
- Added "address lookup table" program
- Number of table entries is capped at `256` since transaction v2 account indices are typed as `u8`
- Lookup tables are append only, and new addresses are not usable until the following block
- Support closing address lookup table accounts but use derived addresses based on a recent slot to prevent new lookup tables from being re-initialized at the same address.

Implements the amended proposal here: https://github.com/solana-labs/solana/pull/21576

### CODE REVIEW

**Comment by jstarry** [comment]
@lijunwangs can you please advise at what the impact of changing these Postgres types and tables will be?

File: programs/address-lookup-table/src/processor.rs [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        if lookup_table.addresses.is_empty() {
+            ic_msg!(invoke_context, "Empty lookup tables cannot be frozen");
+            return Err(InstructionError::InvalidInstructionData);
+        }
+
+        let mut lookup_table_meta = lookup_table.meta;
+        drop(lookup_table_account_ref);
+
+        lookup_table_meta.authority = None;
+        lookup_table_account.set_state(&ProgramState::LookupTable(lookup_table_meta))?;
```
Comment (joncinque) [inline programs/address-lookup-table/src/processor.rs:199]: nit: this is pedantic, but what do you think about overwriting the rest of the meta portion to `0`s after the authority is set to `None`?  Just in case someone uses a bad deserializer where they only look at those bytes, they could get tricked into thinking there's still an authority.  far-fetched? absolutely. but who knows?

File: programs/address-lookup-table/src/processor.rs [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+        drop(lookup_table_account_ref);
+
+        let withdrawn_lamports = lookup_table_account.lamports()?;
+        recipient_account
+            .try_account_ref_mut()?
+            .checked_add_lamports(withdrawn_lamports)?;
+
+        let mut lookup_table_account = lookup_table_account.try_account_ref_mut()?;
+        lookup_table_account.set_data(Vec::new());
```
Comment (joncinque) [inline programs/address-lookup-table/src/processor.rs:372]: :tada:

File: programs/address-lookup-table/src/processor.rs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        }
+
+        // Assert that the slot used in the derivation path of the lookup table address
+        // is no longer recent and can't be reused to initialize an account at the same address.
+        let slot_hashes: SlotHashes = invoke_context.get_sysvar(&slot_hashes::id())?;
+        if let Some(position) = slot_hashes.position(&lookup_table.meta.derivation_slot) {
+            let expiration = MAX_ENTRIES.saturating_sub(position);
+            ic_msg!(
+                invoke_context,
+                "Table cannot be closed until its derivation slot expires in {} blocks",
```
Comment (joncinque) [inline programs/address-lookup-table/src/processor.rs:358]: Should this be?
```suggestion
                "Table cannot be closed until its derivation slot expires in {} slots",
```

File: programs/address-lookup-table-tests/tests/common.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        instruction::InstructionError,
+        pubkey::Pubkey,
+        signature::{Keypair, Signer},
+        slot_hashes::SlotHashes,
+        transaction::{Transaction, TransactionError},
+    },
+    std::borrow::Cow,
+};
+
+#[allow(dead_code)]
```
Comment (joncinque) [inline programs/address-lookup-table-tests/tests/common.rs:22]: it's so annoying to have to put these everywhere :disappointed:

File: programs/address-lookup-table-tests/tests/common.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+pub async fn setup_test_context() -> ProgramTestContext {
+    let program_test = ProgramTest::new("", id(), Some(process_instruction));
```
Comment (joncinque) [inline programs/address-lookup-table-tests/tests/common.rs:24]: interesting, so "native mode" in program-test works for native programs. good to know!

**Review by joncinque — COMMENTED** [review]
As with the first round of this program, looks really good! Only one potentially substantive comment, then this is definitely ready to go

File: program-test/src/lib.rs [hunk 6]
```diff
@@ -1045,6 +1045,18 @@ impl ProgramTestContext {
         bank.store_account(address, account);
     }
 
+    /// Create or overwrite a sysvar, subverting normal runtime checks.
+    ///
+    /// This method exists to make it easier to set up artificial situations
+    /// that would be difficult to replicate on a new test cluster. Beware
+    /// that it can be used to create states that would not be reachable
+    /// under normal conditions!
+    pub fn set_sysvar<T: SysvarId + Sysvar>(&self, sysvar: &T) {
```
[the remark this answers is not in the document]
Comment (jstarry) [reply program-test/src/lib.rs:1054]: Yeah, and it's quite nice that even though new blockhashes are created with program test, it doesn't actually create new banks so sysvars stay in whatever state you set them to!

File: programs/address-lookup-table-tests/tests/common.rs (hunk 5 above)
Comment (jstarry) [reply programs/address-lookup-table-tests/tests/common.rs:24]: Indeed it does! There could be some weird edge cases still but I think we can officially move away from all the nasty mock stuff that's used for testing the vote program, for example.

File: programs/address-lookup-table-tests/tests/common.rs (hunk 4 above)
Comment (jstarry) [reply programs/address-lookup-table-tests/tests/common.rs:22]: Changed to a module-wide allow, which is slightly less annoying haha

File: programs/address-lookup-table/src/instruction.rs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    ///   0. `[WRITE]` Address lookup table account to close
+    ///   1. `[SIGNER]` Current authority
+    ///   2. `[WRITE]` Recipient of closed account lamports
+    CloseLookupTable,
+}
+
+/// Derives the address of an address table account from a wallet address and a recent block's slot.
+pub fn derive_lookup_table_address(
+    authority_address: &Pubkey,
+    recent_block_slot: Epoch,
```
[the remark this answers is not in the document]
Comment (jstarry) [reply programs/address-lookup-table/src/instruction.rs:62]: Ah yeah, just a type alias instead of a new-type, thanks!

File: programs/address-lookup-table/src/processor.rs (hunk 1 above)
Comment (jstarry) [reply programs/address-lookup-table/src/processor.rs:199]: I think that's good practice, will do!

File: programs/address-lookup-table/src/processor.rs (hunk 3 above)
Comment (jstarry) [reply programs/address-lookup-table/src/processor.rs:358]: It's actually blocks. The recent slot hashes sysvar is only updated for new blocks (https://github.com/solana-labs/solana/blob/622fd7c7ecc8d0756a35e911e30ee5f8af4dceb4/runtime/src/bank.rs#L1932)

File: programs/address-lookup-table/src/processor.rs (hunk 2 above)
Comment (jstarry) [reply programs/address-lookup-table/src/processor.rs:372]: Feels good man

File: programs/address-lookup-table/src/state.rs [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    solana_frozen_abi_macro::{AbiEnumVisitor, AbiExample},
+    solana_sdk::{clock::Slot, instruction::InstructionError, pubkey::Pubkey},
+    std::borrow::Cow,
+};
+
+/// The maximum number of addresses that a lookup table can hold
+pub const LOOKUP_TABLE_MAX_ADDRESSES: usize = 256;
+
+/// The serialized size of lookup table metadata
+pub const LOOKUP_TABLE_META_SIZE: usize = 56;
```
[the remark this answers is not in the document]
Comment (jstarry) [reply programs/address-lookup-table/src/state.rs:12]: Yeah it's to 8-byte align addresses which I believe is more efficient somehow. I have fairly low confidence in that but the extra two bytes is ok. I made it clearer with an additional `_padding` field.

File: programs/address-lookup-table/src/state.rs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+#[derive(Debug, PartialEq, Clone, AbiExample)]
+pub struct AddressLookupTable<'a> {
+    pub meta: LookupTableMeta,
+    pub addresses: Cow<'a, [Pubkey]>,
+}
+
+impl<'a> AddressLookupTable<'a> {
+    /// Serialize an address table including its addresses
+    pub fn serialize(&self, mut data: &mut Vec<u8>) -> Result<(), InstructionError> {
+        bincode::serialize_into(&mut data, &(1u32, &self.meta)).map_err(|_| {
```
[the remark this answers is not in the document]
Comment (jstarry) [reply programs/address-lookup-table/src/state.rs:64]: Done

File: programs/address-lookup-table/src/processor.rs [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+            AddressLookupTable::overwrite_meta_data(
+                lookup_table_account_ref_mut.data_as_mut_slice(),
+                lookup_table_meta,
+            )?;
+
+            let table_data = lookup_table_account_ref_mut.data_mut();
+            for new_address in new_addresses {
+                table_data.extend_from_slice(new_address.as_ref());
+            }
+        }
```
Comment (jstarry) [inline programs/address-lookup-table/src/processor.rs:304]: This block changed in the latest commit. Rather than allocating a new account data vector, this tries to modify the `AccountSharedData` `data` vector directly if it isn't already referenced somewhere.

File: programs/address-lookup-table/src/processor.rs (hunk 3 above)
Comment (joncinque) [reply programs/address-lookup-table/src/processor.rs:369]: Gotcha, I had a feeling you were right on this one :smile:

**Review by joncinque — APPROVED** [review]
Not sure if you want anyone else to review, but it's good from my side!
